geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/GeoMesaParam.scala

```diff
@@ -0,0 +1,99 @@
+
+package org.locationtech.geomesa.utils.geotools
+
+import java.io.{IOException, Serializable}
+import java.util.Collections
+
+import com.typesafe.scalalogging.LazyLogging
+import org.geotools.data.DataAccessFactory.Param
+import org.locationtech.geomesa.utils.conf.GeoMesaSystemProperties.SystemProperty
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+import scala.reflect.ClassTag
+
+class GeoMesaParam[T <: AnyRef](_key: String, // can't override final 'key' field from Param
+                                description: String = "",
+                                required: Boolean = false,
+                                val default: T = null,
+                                metadata: Map[String, Any] = null,
+                                val deprecated: Seq[String] = Seq.empty,
+                                val systemProperty: Option[(SystemProperty, (String) => T)] = None)
+                               (implicit ct: ClassTag[T])
+    extends Param(_key, ct.runtimeClass, description, required, default, metadata.asJava) with LazyLogging {
+
+  // ensure that sys property default is the same as param default, otherwise param default will not be used
+  assert(systemProperty.forall { case (prop, convert) => prop.default == null || convert(prop.default) == default })
+
+  /**
+    * Checks that the parameter is contained in the map, but does not do type conversion
+    *
+    * @param params parameter map
+    * @return
+    */
+  def exists(params: java.util.Map[String, _ <: Serializable]): Boolean =
+    params.containsKey(key) || deprecated.exists(params.containsKey)
+
+  /**
+    * Returns the typed value from the map. Priority is:
+    *
+    *   1. primary key from the map
+    *   2. deprecated keys from the map
+    *   3. system property, if defined
+    *   4. default value, if defined
+    *   5. null
+    *
+    * Required parameters must be contained in the map, they will not fall back to
+    * default values or system properties
+    *
+    * @param params parameter map
+    * @return
+    */
+  def lookup(params: java.util.Map[String, _ <: Serializable]): T = {
+    if (params.containsKey(key)) {
+      lookUp(params).asInstanceOf[T]
+    } else if (deprecated.exists(params.containsKey)) {
+      val oldKey = deprecated.find(params.containsKey).get
+      deprecationWarning(oldKey)
+      lookUp(Collections.singletonMap(key, params.get(oldKey))).asInstanceOf[T]
+    } else if (required) {
+      throw new IOException(s"Parameter $key is required: $description")
+    } else {
+      systemProperty.flatMap { case (prop, convert) => prop.option.map(convert) }.getOrElse(default)
+    }
+  }
+
+  /**
+    * Lookup for optional parameters. If a default or system property is defined, will be returned
+    *
+    * @param params parameter map
+    * @return
+    */
+  def lookupOpt(params: java.util.Map[String, Serializable]): Option[T] = Option(lookup(params))
```

Comment:
Is it just convention that this would only be called with optional parameters?  Probably not worth enforcing in any way.

Reply:
yeah, even with optional parameters they have a default it will be returned here. the other option would be to split out an OptionalGeoMesaParam or something like that, although that might be confusing too since if they have a default value they will still never return null.